llvm/lib/CodeGen/CodeGenPrepare.cpp

```diff
@@ -2686,6 +2687,51 @@ bool CodeGenPrepare::optimizeCallInst(CallInst *CI, ModifyDT &ModifiedDT) {
     return true;
   }
 
+  // SCCP may have propagated, among other things, C++ static variables across
+  // calls. If this happens to be the case, we may want to undo it in order to
+  // avoid redundant pointer computation of the constant, as the function method
+  // returning the constant needs to be executed anyways.
+  auto GetUniformReturnValue = [](const Function *F) -> Constant * {
+    if (!F->getReturnType()->isPointerTy())
+      return nullptr;
+
+    Constant *UniformValue = nullptr;
+    for (auto &BB : *F) {
+      if (auto *RI = dyn_cast<ReturnInst>(BB.getTerminator())) {
+        if (auto *V = dyn_cast<GlobalVariable>(RI->getReturnValue())) {
+          if (!UniformValue)
+            UniformValue = V;
+          else if (V != UniformValue)
+            return nullptr;
+        } else {
+          return nullptr;
+        }
+      }
+    }
+
+    return UniformValue;
+  };
+
+  if (!Callee->isInterposable()) {
```

Comment:
```suggestion
  if (!Callee->hasExactDefinition()) {
```
I think we should check for an exact definition here, to avoid any derefinement issues. SCCP will not propagate the return value without exact definition anyway.

(Consider a function that returns a global or poison and the visible definition folds to the global, while an ODR equivalent returns poison on one path.)

Reply:
You can test this using a linkonce_odr function.

Reply:
Agree, thanks, was indeed previously dubious on this as available_externally was not contemplated by isInterposable.